Add headers for multiple language identification

mbatchkarov opened the pull request:
*Issue #, if available:* N/A

*Description of changes:*
Hi, Transcribe SDE here. We recently launched a new feature called [multiple language identification](https://docs.aws.amazon.com/transcribe/latest/dg/lang-id-stream.html#multi-language-streaming). We've been asked to contribute to this package to enable the feature so customers can use it.

*Notes*
`language_code` is currently a required positional parameter. When language ID is added, language code should become optional. That means we'd have to make it the third param and give it a default value, but this would break existing clients that use positional-only arguments. Therefore I'm leaving it as a positional arg and requiring that it's set to None when language ID is enabled. I'm not too happy with this option either- happy to discuss. Maybe something like this would be more ergonomic:

```
        if identify_language or identify_language:
            warnings.warm("Setting language_code to None because language ID is enabled")
            language_code = None
```


*Testing*
I extended and ran the integration tests locally from my own AWS account. 


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

mbatchkarov committed:
add headers for multiple language identification

mbatchkarov commented:
Update: just saw #89 - looks like plain language ID is also not supported. Will update the PR to include that too

mbatchkarov committed:
add headers for language identification

mbatchkarov commented:
Update: just saw #89 - looks like plain language ID is also not supported. Will update the PR to include that too

praveenXira commented:
When will this PR get merged? I really need [multiple language identification](https://docs.aws.amazon.com/transcribe/latest/dg/lang-id-stream.html#multi-language-streaming) feature.

GameSetAndMatch commented on amazon_transcribe/client.py:
```diff
@@ -84,6 +84,10 @@ async def start_stream_transcription(
+        identify_language: Optional[bool] = False,
+        preferred_language: Optional[str] = None,
+        identify_multiple_languages=False,
+        language_options=None,
```
I am not familiar with AWS coding standards, but is there a reason why there isn't type hiting on `identify_multiple_languages` and `language_options` ?

pencil replied:
```suggestion
        language_options: Optional[List[str]] = None,
```

GameSetAndMatch commented:
Definitely a feature I would use in the short term if it was reviewed and merged by AWS team, good work @mbatchkarov !

gunwooterry commented on amazon_transcribe/serialize.py:
```diff
@@ -130,6 +134,36 @@ def serialize_start_stream_transcription_request(
+
+        headers.update(
+            self._serialize_str_header(
+                "preferred_language",
```
```suggestion
                "preferred-language",
```
I think it's a typo.

mbatchkarov committed:
Fix typo in header name

Header names are separated by hyphens, not underscores

https://docs.aws.amazon.com/transcribe/latest/dg/lang-id-stream.html

Co-authored-by: Gunwoo Kim <gunwooterry@gmail.com>

pencil commented on amazon_transcribe/client.py:
```diff
@@ -84,6 +84,10 @@ async def start_stream_transcription(
         language_model_name: Optional[str] = None,
+        identify_language: Optional[bool] = False,
+        preferred_language: Optional[str] = None,
+        identify_multiple_languages=False,
```
```suggestion
        identify_multiple_languages: Optional[bool] = None,
```

pencil commented on amazon_transcribe/client.py:
```diff
@@ -84,6 +84,10 @@ async def start_stream_transcription(
         enable_partial_results_stabilization: Optional[bool] = None,
         partial_results_stability: Optional[str] = None,
         language_model_name: Optional[str] = None,
+        identify_language: Optional[bool] = False,
```
```suggestion
        identify_language: Optional[bool] = None,
```

pencil reviewed (commented):
Thanks for putting this up! Unfortunately, it looks like AWS has abandoned this client 😒

kandakji commented:
why isn't this already merged?

mbatchkarov committed:
address CR comments and add language_code in response

mirobat reviewed (dismissed):
TODO- add support for CVs

mbatchkarov committed:
add support for CV when using Multi LID

mbatchkarov committed:
remove python 3.7 from test matrix

it reached end of life in mind 2023 and github runnners no longer seem to support it

mbatchkarov committed:
remove all uses of py 3.7 and replace with 3.11

mbatchkarov committed:
update codecov action

the run is failing https://github.com/awslabs/amazon-transcribe-streaming-sdk/actions/runs/13289281664/job/37105650123 - need to remove the yml attribute, and update to the latest action version while we're at it

mbatchkarov committed:
fix black warning

mbatchkarov committed:
update deprecated ubuntu used in linter

mbatchkarov committed:
exclude tests from coverage

mbatchkarov committed:
fix flake8 warning

xuejiey commented on amazon_transcribe/client.py:
```diff
@@ -144,12 +153,29 @@ async def start_stream_transcription(
+            You must also provide at least two language_options and set
+            language_code to None
+        :param language_options:
+            A list of possible language to use when identify_multiple_languages is
```
Should mention identify_language as well

xuejiey commented on amazon_transcribe/client.py:
```diff
@@ -100,14 +105,18 @@ async def start_stream_transcription(
         :param language_code:
-            Indicates the source language used in the input audio stream.
+            Indicates the source language used in the input audio stream. Set to
+            None if identify_multiple_languages is set to True
```
Should mention identify_language as well

xuejiey commented on amazon_transcribe/model.py:
```diff
@@ -226,6 +229,15 @@ class StartStreamTranscriptionRequest:
+            You must also provide at least two language_options and set
+            language_code to None
+    : param language_options:
+        A list of possible language to use when identify_multiple_languages is
```
Should mention identify_language as well
NIT: Space after `:`

xuejiey commented on amazon_transcribe/client.py:
```diff
@@ -75,6 +75,7 @@ async def start_stream_transcription(
+        vocabulary_names: Optional[List[str]] = None,
         session_id: Optional[str] = None,
         vocab_filter_method: Optional[str] = None,
         vocab_filter_name: Optional[str] = None,
```
Can you add `VocabularyFilterNames`

mbatchkarov committed:
add support for vocabulary filters

mbatchkarov committed:
fix bad merge :(

pencil commented:
@xuejiey @mirobat  Thanks for getting this shipped! It's great to see AWS working on this client again <3
